src/clients/python/src/tigerbeetle_python/client.py

```diff
@@ -0,0 +1,229 @@
+from __future__ import annotations
+
+import asyncio
+import ctypes
+import logging
+import os
+import threading
+import time
+from collections.abc import Callable  # noqa: TCH003
+from dataclasses import dataclass
+from typing import Any
+
+from . import bindings
+
+logger = logging.getLogger("tigerbeetle")
+
+
+class AtomicInteger:
+    def __init__(self, value=0):
+        self._value = value
+        self._lock = threading.Lock()
+
+    def increment(self):
+        with self._lock:
+            self._value += 1
+            return self._value
+
+
+@dataclass
+class CompletionContextSync:
+    event: threading.Event
+
+
+@dataclass
+class CompletionContextAsync:
+    loop: asyncio.AbstractEventLoop
+    event: asyncio.Event
+
+
+@dataclass
+class InflightPacket:
+    packet: bindings.CPacket
+    response: Any
+    operation: bindings.Operation
+    c_event_type: Any
+    c_result_type: Any
+    on_completion: Callable | None
+    on_completion_context: CompletionContextSync | CompletionContextAsync | None
+
+
+
+def id() -> int:
+    """
+    Generates a Universally Unique and Sortable Identifier as a 128-bit integer. Based on ULIDs.
+    """
+    time_ms = time.time_ns() // (1000 * 1000)
+
+    # Ensure time_ms monotonically increases.
+    time_ms_last = getattr(id, "_time_ms_last", 0)
+    if time_ms <= time_ms_last:
+        time_ms = time_ms_last
+    else:
+        id._time_ms_last = time_ms
+
+    randomness = os.urandom(10)
+
+    return int.from_bytes(
+        time_ms.to_bytes(6, "big") + randomness,
+        "big",
+    )
+
+
+amount_max = (2 ** 128) - 1
+
+
+class InitError(Exception):
+    pass
+
+class PacketError(Exception):
+    pass
+
+
+class Client:
+    _clients: dict[int, Any] = {}
+    _counter = AtomicInteger()
+
+    def __init__(self, cluster_id: int, replica_addresses: str):
+        self._client_key = Client._counter.increment()
+        self._client = bindings.Client()
+
+        self._inflight_packets: dict[int, InflightPacket] = {}
+
+        init_status = bindings.tb_client_init(
+            ctypes.byref(self._client),
+            cluster_id,
+            replica_addresses.encode("ascii"),
+            len(replica_addresses),
+            self._client_key,
+            self._c_on_completion
+        )
+        if init_status != bindings.Status.SUCCESS:
+            raise InitError(init_status)
+
+        Client._clients[self._client_key] = self
+
+
+    def _acquire_packet(self, operation: bindings.Operation, operations: Any,
+                        c_event_type: Any, c_result_type: Any) -> InflightPacket:
+        packet = bindings.CPacket()
+        packet.next = None
+        packet.user_data = Client._counter.increment()
+        packet.operation = operation
+        packet.status = bindings.PacketStatus.OK
+
+        operations_array_type = c_event_type * len(operations)
+        operations_array = operations_array_type(*map(c_event_type.from_param, operations))
```

Comment:
Hm, I am not sure about the memory management story. Here, operations_array owns the memory we passed to C. When we call `ctypes.cast`, would that incerement the ref-count, such that, when `operations_array` goes out of scope, the underlying memory is still tracked by `packed` through `packet.data` field?

Reply:
I guess, if it doesn't leak and doesn't crash that's fine, but still want to double-check.

Reply:
> When we call ctypes.cast, would that incerement the ref-count, such that, when operations_array goes out of scope, the underlying memory is still tracked by packed through packet.data field?

It does!

```python
import sys
from ctypes import *

operations_array_type = c_uint32 * 32
x = operations_array_type(*range(0, 32))

sys.getrefcount(x) # prints 2

foo = cast(x, c_void_p)
sys.getrefcount(x) # prints 3

```